### PR TITLE
feat: Findex 5.0.0 + Findex Redis impl

### PR DESCRIPTION
**Pull request by bgrieder**
This PR mainly 

 - upgrades Findex to 5.0.0
 -  adds  a full (i.e. including `compact()`) implementation of Findex on Redis
 
 Other minor modifications
 - better separate tests from core features of the (partial) implementation of Findex over Sqlite
 - adds PyO3 build support on macOS

⚠️ the main Cargo.toml has a relative path to `cloudproof_findex`

### CODE REVIEW

File: crates/findex/src/implementations/redis/findex.rs [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    }
+
+    async fn fetch_all_entry_table_uids(&self) -> Result<Uids<UID_LENGTH>, FindexRedisError> {
+        let keys: Vec<Vec<u8>> = self.mgr.clone().keys(key(FindexTable::Entry, b"*")).await?;
+        trace!("fetch_all_entry_table_uids num keywords: {}", keys.len());
+        Ok(Uids(
+            keys.iter()
+                .map(|v| {
+                    let mut uid = [0u8; UID_LENGTH];
+                    uid.copy_from_slice(&v[4..]);
```
Comment (ThibaudDauce) [inline crates/findex/src/implementations/redis/findex.rs:217]: The `4` here is a little magic, maybe store `b"fe::"` inside a constant and use the length of this constant instead?

File: crates/findex/src/implementations/redis/findex.rs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+        Ok(EncryptedTable::from(
+            chain_table_uids
+                .0
+                .into_iter()
+                .zip(values)
+                .filter(|(_uid, v)| !v.is_empty())
+                .collect::<HashMap<Uid<UID_LENGTH>, Vec<u8>>>(),
+        ))
+    }
```
Comment (ThibaudDauce) [inline crates/findex/src/implementations/redis/findex.rs:288]: Duplicate code, we could have a `fetch` function inside the `FindexRedis` implementation (outside the `FindexCallbacks` impl) which takes UIDs and `FindexTable` and call it.

Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:288]: Since these functions do not return the same type, the amount of deduplication will be very small at the expense of readability. 
Keeping as such

Comment (ThibaudDauce) [reply crates/findex/src/implementations/redis/findex.rs:288]: Yes good point! Didn't remember the change about return types :-(

File: crates/findex/src/implementations/redis/findex.rs [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        &self,
+        chain_table_uids_to_remove: Uids<UID_LENGTH>,
+        new_encrypted_entry_table_items: EncryptedTable<UID_LENGTH>,
+        new_encrypted_chain_table_items: EncryptedTable<UID_LENGTH>,
+    ) -> Result<(), FindexRedisError> {
+        // Prevent compacting while already compacting
+        // We could add support to make sure we are not upserting
+        if self.compact_lock.try_read().is_err() {
+            return Err(FindexRedisError::Compacting);
+        }
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:340]: Same here, I think the problem is only partially solved. Another action needs to be taken anyway so I would remove this.

File: crates/findex/src/implementations/redis/findex.rs [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+            new_encrypted_entry_table_items.len(),
+            new_encrypted_chain_table_items.len()
+        );
+        // delete the entry table
+        let entry_keys: Vec<Vec<u8>> = self.mgr.clone().keys(key(FindexTable::Entry, b"*")).await?;
+        let mut pipeline = pipe();
+        for entry_key in entry_keys {
+            pipeline.del(entry_key);
+        }
+        pipeline.atomic().query_async(&mut self.mgr.clone()).await?;
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:356]: If your client crashes after this, you end up losing all your index. It is best to delete old ET lines last. That way clients still can search and upsert the old index.

File: crates/findex/src/implementations/redis/findex.rs [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        trace!("upsert_entry_table num keywords {:?}", modifications.len());
+
+        let mut rejected = EncryptedTable::default();
+        for (uid, (old_value, new_value)) in modifications {
+            let value: Vec<u8> = self
+                .upsert_script
+                .arg(key(FindexTable::Entry, &uid))
+                .arg(old_value.unwrap_or_default())
+                .arg(new_value)
+                .invoke_async(&mut self.mgr.clone())
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:308]: `self.mgr` seems to be a non-trivial object. Have you considered using an `Arc` instead of cloning it around?

File: crates/findex/src/implementations/redis/findex.rs [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+    ///
+    /// - `master_key`  : Findex master key
+    /// - `label`       : additional public information used in key hashing
+    /// - `additions`   : values to indexed for a set of keywords
+    /// - `deletions`   : values to remove from the indexes for a set of
+    ///   keywords
+    ///
+    /// # Returns
+    /// The set of new keywords that were added
+    pub async fn upsert(
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:117]: Can you explain the rational behind these Findex API wrappers?

File: crates/findex/src/lib.rs [hunk 7]
```diff
@@ -5,6 +5,7 @@
 #![cfg_attr(any(feature = "wasm_bindgen", feature = "ffi"), feature(const_option))]
 #![cfg_attr(feature = "cloud", feature(iter_next_chunk))]
 
+pub mod implementations;
```
Comment (tbrezot) [inline crates/findex/src/lib.rs:8]: You may hide this declaration behind the features `sqlite` or `redis`.

File: crates/findex/tests/sqlite/database.rs [hunk 8]
```diff
@@ -1,3 +1,10 @@
+//! This file contains the implementation of the Sqlite database used for the
+//! tests.
+#![allow(dead_code)]
```
Comment (tbrezot) [inline crates/findex/tests/sqlite/database.rs:3]: You shouldn't need this.
```suggestion
```

File: crates/findex/src/implementations/redis/findex.rs [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+
+/// Generate a key for the entry table or chain table
+fn key(table: FindexTable, uid: &[u8]) -> Vec<u8> {
+    let mut key = match table {
+        FindexTable::Entry => b"fe::".to_vec(),
+        FindexTable::Chain => b"fc::".to_vec(),
+    };
+    key.extend_from_slice(uid);
+    key
+}
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crates/findex/src/implementations/redis/findex.rs:33]: Yes, ET and CT are stored in the same Redis (Redis doesn't have the concept of table).

In Redis you need to have a prefix to split the keys for `fetch_all_entry_table_uids` you cannot add a boolean to the value and filter on it. But the prefix could be smaller (I'm not sure it's worth it)

File: crates/findex/src/implementations/redis/findex.rs [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+        // mget the values from the Redis keys
+        let values: Vec<Vec<u8>> = self.mgr.clone().mget(keys).await?;
+
+        // discard empty values
+        let tuples = entry_table_uids
+            .0
+            .into_iter()
+            .zip(values)
+            .filter(|(_uid, v)| !v.is_empty())
+            .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crates/findex/src/implementations/redis/findex.rs:253]: I think the iterator is in arbitrary order but constant in the same run. So it's fine here (the same iteration is done above to create the Redis keys and Redis returns the values in the same order, so the second iteration should be in the same order as the first above)

File: crates/findex/src/implementations/redis/findex.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs:33]: I think this is worth modifying:
- 4 characters -> 12.5% overhead on UID storage -> ~4% overhead on the Entry Table and 2/3% overhead on the Chain Table
- the copy of each UID in a newly allocated vector wastes time and memory

I think it may be possible to simply set the first bit of each UID to 1 or 0. No overhead, no extra allocation and the UID size guarantee there will be no collision. I will double-check with @chloehebant.

File: crates/findex/src/implementations/redis/findex.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs:253]: Yes, as long as it does not change between two cals. My point is that it is the case now, but there is no guarantee that this will be true in the future. But we still can change this piece of code the day it does not work anymore so this is okay.

There is still an issue with the multiple ET though. Do we ignore this use case for the Redis callback ?

[the remark this answers is not in the document]
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs:253]: This code also relies on the assumption that the call to Redis does not shuffle the UIDs.

At the very least, these two implicit assumptions should be documented in a comment imho.

File: crates/findex/src/implementations/redis/findex.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:33]: The Redis server can be used to store other things (in the KMS, the KMIP objects for instance), so I would like to keep the prefix discriminant enough but have reduced it to size 2

The code is now

```
/// The length of the prefix of the table name in bytes
/// 0x00ee for the entry table
/// 0x00ef for the chain table
const TABLE_PREFIX_LENGTH: usize = 2;

#[derive(Copy, Clone)]
enum FindexTable {
    Entry = 0xee,
    Chain = 0xef,
}

/// Generate a key for the entry table or chain table
fn key(table: FindexTable, uid: &[u8]) -> Vec<u8> {
    [&[0x00, table as u8], uid].concat()
}
```

File: crates/findex/src/implementations/redis/findex.rs [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+        FindexTable::Entry => b"fe::".to_vec(),
+        FindexTable::Chain => b"fc::".to_vec(),
+    };
+    key.extend_from_slice(uid);
+    key
+}
+
+pub struct FindexRedis {
+    // we keep redis_url for the updateLines method
+    mgr: ConnectionManager,
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:37]: ok

File: crates/findex/src/implementations/redis/findex.rs (hunk 1 above)
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:217]: Now in a constant 

```
/// The length of the prefix of the table name in bytes
/// 0x00ee for the entry table
/// 0x00ef for the chain table
const TABLE_PREFIX_LENGTH: usize = 2;
```

File: crates/findex/src/implementations/redis/findex.rs [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+        chain_table_uids_to_remove: Uids<UID_LENGTH>,
+        new_encrypted_entry_table_items: EncryptedTable<UID_LENGTH>,
+        new_encrypted_chain_table_items: EncryptedTable<UID_LENGTH>,
+    ) -> Result<(), FindexRedisError> {
+        // Prevent compacting while already compacting
+        // We could add support to make sure we are not upserting
+        if self.compact_lock.try_read().is_err() {
+            return Err(FindexRedisError::Compacting);
+        }
+        let _compact_lock = self.compact_lock.write().await;
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:341]: yes 

```
let _compact_lock = match self.compact_lock.try_write() {
            Ok(lock) => lock,
            Err(_) => return Err(FindexRedisError::Compacting), /* another thread is compacting already so we return an error */
        };
 ```

File: crates/findex/tests/sqlite/database.rs (hunk 8 above)
Comment (bgrieder) [reply crates/findex/tests/sqlite/database.rs:3]: yes

File: crates/findex/src/implementations/redis/findex.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:253]: Redis does not shuffle by design.

For the rest, not sure there is a problem but to make sure (at the cost of another iterator)

```
       // collect uids from the entry table
        let uids: Vec<Uid<32>> = entry_table_uids.0.into_iter().collect();

        // build Redis keys
        let keys: Vec<Vec<u8>> = uids
            .iter()
            .map(|uid| key(FindexTable::Entry, uid))
            .collect();

        // mget the values from the Redis keys
        let values: Vec<Vec<u8>> = self.manager.clone().mget(keys).await?;

        // discard empty values
        let tuples = uids
            .into_iter()
            .zip(values)
            .filter(|(_uid, v)| !v.is_empty())
            .collect::<Vec<_>>();
```

Same fix reported on the chain table

File: crates/findex/src/implementations/redis/findex.rs (hunk 5 above)
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:308]: This is the doc recommended usage

File: crates/findex/src/implementations/redis/findex.rs (hunk 6 above)
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:117]: Yes, it is a common pattern called a "facade". It does two things:
 - avoid the user importing traits e.g., `FindexSearch`, `FindexUpdate`, etc...
 - avoid exposing the user to public methods in those traits that it should not be using
 
 As a general rule, implementations should always provide facades to complex APIs, to simplify the user API

File: crates/findex/src/implementations/redis/findex.rs [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+    }
+
+    async fn upsert_entry_table(
+        &self,
+        modifications: UpsertData<UID_LENGTH>,
+    ) -> Result<EncryptedTable<UID_LENGTH>, FindexRedisError> {
+        // Prevent upserting while compacting
+        if self.compact_lock.try_read().is_err() {
+            return Err(FindexRedisError::Compacting);
+        }
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:297]: No, it does not. But at least it prevents the issue when there is a single multi-threaded client (such as the KMS)

File: crates/findex/src/implementations/redis/findex.rs [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+                .upsert_script
+                .arg(key(FindexTable::Entry, &uid))
+                .arg(old_value.unwrap_or_default())
+                .arg(new_value)
+                .invoke_async(&mut self.mgr.clone())
+                .await?;
+            if !value.is_empty() {
+                rejected.insert(uid, value);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:313]: not with the script

File: crates/findex/src/implementations/redis/findex.rs (hunk 3 above)
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:340]: see comment above

File: crates/findex/src/lib.rs (hunk 7 above)
Comment (bgrieder) [reply crates/findex/src/lib.rs:8]: ok

File: crates/findex/src/implementations/redis/findex.rs (hunk 4 above)
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:356]: Some new keys can be identical to old ones, so it is tricky; it also slows down stuff considerably on large compacts, but yes, I implemented it

File: crates/findex/src/implementations/redis/findex.rs [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+    ///
+    /// # Parameters
+    ///
+    /// - `master_key`          : Findex master key
+    /// - `label`               : public label
+    /// - `keywords`            : keywords to search
+    /// - `max_depth`           : maximum recursion depth allowed
+    pub async fn search(
+        &self,
+        master_key: &[u8; 16],
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crates/findex/src/implementations/redis/findex.rs:146]: ok

**Comment by bgrieder** [comment]
OK. Review done. 

Anythoing else ?

File: Cargo.toml [hunk 16]
```diff
@@ -1,26 +1,31 @@
...
 # some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`. Fix the resolver value to avoid surprises.
 resolver = "1"
 
 [workspace.dependencies]
 async-trait = "0.1.51"
 base64 = "0.21.0"
-cosmian_crypto_core = "9.0.0"
-cloudproof_cover_crypt = "12.0.1"
+cosmian_crypto_core = "9.0"
+cloudproof_cover_crypt = "12.0"
```
Comment (tbrezot) [inline Cargo.toml]: I thought we settled on using the full semver number since it prevents from using earlier patch releases while allowing to upgrad to the next patch/minor release.

Why did you remove the patch number ?

File: crates/findex/src/implementations/redis/findex.rs [hunk 17]
```diff
@@ -0,0 +1,466 @@
...
+            removed_locations_finder,
+            compact_lock: RwLock::new(()),
+        })
+    }
+
+    /// Connect to a Redis server with a `ConnectionManager`
+    ///
+    /// # Arguments
+    ///  * `redis_url` - The Redis URL e.g.
+    ///    "redis://user:password@localhost:6379"
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:81]:
```suggestion
```

File: crates/findex/src/implementations/redis/findex.rs [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+            // do not delete the new keys
+            if new_keys.contains(&entry_key) {
+                continue;
+            }
+            pipeline.del(entry_key);
+        }
+        pipeline
+            .atomic()
+            .query_async(&mut self.manager.clone())
+            .await?;
```
Comment (tbrezot) [inline crates/findex/src/implementations/redis/findex.rs:409]: The order is still wrong. You should:

1. save old entry UIDs
2. add new chains
3. add new entries
4. publish the new label
5. delete old entries
6. delete old chains

That way, you always maintain a working system in case of failure (cf the [documentation](https://github.com/Cosmian/findex/blob/fce3735f76e7fcb4c2d505205f74ac55b2052407/src/callbacks.rs#L122) of the option 2 of the upsert function in Findex).

**Comment by tbrezot** [comment]
> OK. Review done.
> 
> Anythoing else ?

Thanks.

There is still an issue with the compacting operation (the order of the operations is not correct). The rest of it is just some remarks and propositions.

File: crates/findex/src/implementations/redis/findex.rs [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+    /// Clear all indexes
+    ///
+    /// # Warning
+    /// This is definitive
+    pub async fn clear_indexes(&self) -> Result<(), FindexRedisError> {
+        redis::cmd("FLUSHDB")
+            .query_async(&mut self.manager.clone())
+            .await?;
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs:99]: Rust does not allow exporting test interfaces. This does not work.

File: crates/findex/src/implementations/redis/findex.rs (hunk 18 above)
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs]: Done in 7867bf8

File: crates/findex/src/implementations/redis/findex.rs [hunk 20]
```diff
@@ -0,0 +1,466 @@
...
+        );
+        // collect all the entry table keys to delete
+        let entry_keys_to_delete: Vec<Vec<u8>> = self
+            .manager
+            .clone()
+            .keys(key(FindexTable::Entry, b"*"))
+            .await?;
+
+        // add new entry table entries
+        // keep track of the keys we added so we don't delete them later on
```
[the remark this answers is not in the document]
Comment (tbrezot) [reply crates/findex/src/implementations/redis/findex.rs]: Done in 7867bf8
